### PR TITLE
Add `present` to `file`'s `state` attribute.

### LINKED ISSUE #170: file touch always 'changed' - [was: need a separate touch module]
##### Issue Type:

Feature Idea
##### Ansible Version:

ansible 1.7.2
##### Environment:

N/A
##### Summary:

`file` `touch` is useless, because it always completes with `changed` status
##### Steps To Reproduce:

```
file: path=/var/log/php.log state=touch mode=0666
```
##### Expected Results:

```
touch: path=/var/log/php.log mode=0666 timestamp=no
```
##### Actual Results:

```
# doesn't work correctly, ignores ACL
copy: dest=/var/log/php.log content="" mode=0666 force=no
```

### PR DESCRIPTION
**Pull request by supertylerc**
##### ISSUE TYPE

Feature Pull Request
##### COMPONENT NAME

`file`
##### ANSIBLE VERSION

```
ansible 2.0.1.0
  config file =
  configured module search path = Default w/o overrides
```
##### SUMMARY

Add a `present` value to the `file` module's `state` attribute.  This
allows the `file` module to operate idempotently on files to ensure a
file exists without actually modifying it.

Fixes #170.
###### Before

Using `state: touch`:

```
# First time
       TASK: [example | Touch File] ****************************************
       <localhost> REMOTE_MODULE file group=example state=touch path=/etc/foo.conf owner=example
       changed: [localhost] => {"changed": true, "dest": "/etc/foo.conf", "gid": 1002, "group": "example", "mode": "0600", "owner": "example", "size": 0, "state": "file", "uid": 1002}

# Second time
       TASK: [example | Touch File] ****************************************
       <localhost> REMOTE_MODULE file group=example state=touch path=/etc/foo.conf owner=example
       changed: [localhost] => {"changed": true, "dest": "/etc/foo.conf", "gid": 1002, "group": "example", "mode": "0600", "owner": "example", "size": 0, "state": "file", "uid": 1002}
```
###### After

Using `state: present`

```
# First time
       TASK: [example | Touch File] ****************************************
       <localhost> REMOTE_MODULE file group=example state=present path=/etc/foo.conf owner=example
       changed: [localhost] => {"changed": true, "dest": "/etc/foo.conf", "gid": 1002, "group": "example", "mode": "0600", "owner": "example", "size": 0, "state": "file", "uid": 1002}

# Second time
       TASK: [example | Touch File] ****************************************
       <localhost> REMOTE_MODULE file group=example state=present path=/etc/foo.conf owner=example
       ok: [localhost] => {"changed": false, "dest": "/etc/foo.conf", "gid": 1002, "group": "example", "mode": "0600", "owner": "example", "size": 0, "state": "file", "uid": 1002}
```

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @supertylerc for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by supertylerc** [comment]
Bumpy bump!  @gregdek, is there anything I can do to move things along a bit?

**Comment by gregdek** [comment]
@supertylerc the file module is pretty central, so we really do need the Ansible team to look at it carefully.

From my cursory look, it makes sense and I understand the logic: "present" meaning "touch an empty file" seems reasonable. I suspect there's a reason that this wasn't done in the past, though.

Pinging the @ansible/core team for input here.

File: files/file.py [hunk 1]
```diff
@@ -62,10 +62,12 @@
         directories will be recursively deleted, and files or symlinks will be unlinked.
         If C(touch) (new in 1.4), an empty file will be created if the C(path) does not
         exist, while an existing file or directory will receive updated file access and
-        modification times (similar to the way `touch` works from the command line).
+        modification times (similar to the way `touch` works from the command line).  If
+        C(present), the file will be created if it does not exist.  If it does exist, no
```
Comment (ryansb) [inline files/file.py:66]: "the file will only be created if it does not exist. Otherwise, the modification time will not be changed."

**Comment by alikins** [comment]
pr #3395 , meet pr #3095 . 
3095, meet 3395.

**Comment by gregdek** [comment]
Thanks @supertylerc for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@supertylerc A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@supertylerc Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@supertylerc A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@supertylerc Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@supertylerc A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Pull request closed without merging**